static/app/views/dashboardsV2/widgetCard/genericWidgetQueries.tsx

```diff
@@ -0,0 +1,325 @@
+import {Component} from 'react';
+import isEqual from 'lodash/isEqual';
+
+import {Client, ResponseMeta} from 'sentry/api';
+import {isSelectionEqual} from 'sentry/components/organizations/pageFilters/utils';
+import {t} from 'sentry/locale';
+import {Organization, PageFilters} from 'sentry/types';
+import {Series} from 'sentry/types/echarts';
+import {TableDataRow, TableDataWithTitle} from 'sentry/utils/discover/discoverQuery';
+
+import {DatasetConfig} from '../datasetConfig/base';
+import {DEFAULT_TABLE_LIMIT, DisplayType, Widget, WidgetQuery} from '../types';
+
+function getReferrer(displayType: DisplayType) {
+  let referrer: string = '';
+
+  if (displayType === DisplayType.TABLE) {
+    referrer = 'api.dashboards.tablewidget';
+  } else if (displayType === DisplayType.BIG_NUMBER) {
+    referrer = 'api.dashboards.bignumberwidget';
+  } else if (displayType === DisplayType.WORLD_MAP) {
+    referrer = 'api.dashboards.worldmapwidget';
+  } else {
+    referrer = `api.dashboards.widget.${displayType}-chart`;
+  }
+
+  return referrer;
+}
+
+export type GenericWidgetQueriesChildrenProps = {
+  loading: boolean;
+  errorMessage?: string;
+  pageLinks?: null | string;
+  tableResults?: TableDataWithTitle[];
+  timeseriesResults?: Series[];
+  totalCount?: string;
+};
+
+export type GenericWidgetQueriesProps<SeriesResponse, TableResponse> = {
+  api: Client;
+  children: (props: GenericWidgetQueriesChildrenProps) => JSX.Element;
+  config: DatasetConfig<SeriesResponse, TableResponse>;
+  organization: Organization;
+  selection: PageFilters;
+  widget: Widget;
+  cursor?: string;
+  limit?: number;
+  onDataFetched?: ({
+    tableResults,
+    timeseriesResults,
+    issuesResults,
+    totalIssuesCount,
+    pageLinks,
+  }: {
+    issuesResults?: TableDataRow[];
+    pageLinks?: string;
+    tableResults?: TableDataWithTitle[];
+    timeseriesResults?: Series[];
+    totalIssuesCount?: string;
+  }) => void;
+  processRawSeriesResult?: (result: SeriesResponse) => void;
+  processRawTableResult?: (result: TableResponse) => void;
+};
+
+type State<SeriesResponse> = {
+  loading: boolean;
+  errorMessage?: GenericWidgetQueriesChildrenProps['errorMessage'];
+  pageLinks?: GenericWidgetQueriesChildrenProps['pageLinks'];
+  queryFetchID?: symbol;
+  rawResults?: SeriesResponse[];
+  tableResults?: GenericWidgetQueriesChildrenProps['tableResults'];
+  timeseriesResults?: GenericWidgetQueriesChildrenProps['timeseriesResults'];
+};
+
+class GenericWidgetQueries<SeriesResponse, TableResponse> extends Component<
```

Comment:
In the future would it make sense for this to become just a hook instead?

Reply:
Hmm, good point. I guess it's just a querying component and right now I don't see why that logic can't be contained in a hook and the right properties returned. It wouldn't be until all of its dependencies are converted to functional components that we'd be able to do that, right? We still have to convert a few and see how this works in other use cases.

Reply:
yeah, you're correct. BUT, you could take all of this and put it in a hook, and make a single render props style component that wraps the hook :)